src/optimization/steepest_descent_method.cpp

```diff
@@ -0,0 +1,287 @@
+    , precision_(precision)
+{
+}
+
+SteepestDescentMethod* SteepestDescentMethod::Clone() const
+{
+    return new SteepestDescentMethod(
+        increment_of_the_argument_,
+        max_iteration_number_,
+        precision_);
+}
+
+std::string SteepestDescentMethod::get_name() const
+{
+    return "Steepest Descent Method";
+}
+
+
+#ifdef BUILD_WITH_MPI
+void SteepestDescentMethod::Send(boost::mpi::communicator* communicator, int target) const
+ {
+    communicator->send(target, k52::parallel::mpi::constants::kCommonTag, increment_of_the_argument_);
+    communicator->send(target, k52::parallel::mpi::constants::kCommonTag, max_iteration_number_);
+    communicator->send(target, k52::parallel::mpi::constants::kCommonTag, precision_);
+}
+
+void SteepestDescentMethod::Receive(boost::mpi::communicator* communicator, int source)
+{
+    communicator->recv(source, k52::parallel::mpi::constants::kCommonTag, increment_of_the_argument_);
+    communicator->recv(source, k52::parallel::mpi::constants::kCommonTag, max_iteration_number_);
+    communicator->recv(source, k52::parallel::mpi::constants::kCommonTag, precision_);
+}
+#endif
+
+
+vector<double> SteepestDescentMethod::FindOptimalParameters(const vector<double>& initial_parameters)
+{
+    vector<double> parameters(initial_parameters);
+    size_t counter = 0;
+    double lambda = 1; 
+
+    vector<double> next_step_parameters(parameters);
+    do
+    {
+        parameters = next_step_parameters;
+        counter++;
+        vector<double> gradient = CalculateGradient(parameters);
+        vector<double> invert_gradiend = InvertVector(gradient);
+
+        lambda = GoldenSectionSearch(parameters, lambda, invert_gradiend);
+
+        for (size_t i = 0; i < next_step_parameters.size(); ++i)
+        {
+            next_step_parameters[i] = parameters[i] + lambda * invert_gradiend[i];
+        }
+
+    } while (!IsExitCriteriaFulfilled(next_step_parameters) 
+        && !(counter == max_iteration_number_));
+
+    return parameters;
+}
+
+bool SteepestDescentMethod::IsExitCriteriaFulfilled(
+    const vector<double>& steps_array) const
+{
+    double temp = 0;
+    vector<double> gradient = CalculateGradient(steps_array);
+    for (size_t i = 0; i < steps_array.size(); ++i)
+    {
+        temp += gradient[i] * gradient[i];
+    }
+    if (sqrt(temp) > precision_)
+    {
+        return false;
+    }
+    return true;
+}
+/// @todo: needs to replaced to the method from Math
+vector<double> SteepestDescentMethod::CalculateGradient(
+    const vector<double>& parameters) const
+{
+    vector<double> gradient(parameters.size());
+
+    for (size_t i = 0; i<parameters.size(); ++i)
+    {
+        gradient[i] = CalculateDerivative(parameters, i);
+    }
+
+    return gradient;
+}
+
+/// @todo: needs to replaced to the method from Math
+double SteepestDescentMethod::CalculateDerivative(
+    const vector<double>& parameters,
+    const size_t& index) const
+{
+    double half = 0.5;
```

Comment:
@AlexanderSoroka agreed to fix this for you, Add //TODO comments for 2 copypasted methods with the source of the copypaste. Also create an issue in k52 (if you will have enough privileges) with links to these 2 todo's and assign it to @AlexanderSoroka